Better error logging and faster restarting of dying workers by the supervisor.

knusbaum opened the pull request:
Fix for: https://github.com/nathanmarz/storm/issues/489
(Launching a process throws away standard out; can hang)

nathanmarz commented:
Can you give a high level overview of what this implementation does? From glancing at it, I don't like that it launches a thread per launched worker and the use of global state in the dead worker functions. Any state should be kept in the supervisor-state map.

knusbaum commented:
When a worker process is spawned, a thread is created to log its standard output and then wait for its termination. On termination, the process is marked as dead in the dead-workers structure. 

If the worker isn't supposed to be dead, read-allocated-workers will notice that one of its workers is in the dead-workers structure, log that fact, and immediately attempt to restart the process.

I couldn't find a nice clean way around spawning a thread for each worker process, and since these threads are going to be blocked for the vast majority of their lives, I thought it was acceptable. I did look into making the processes into one thread, but two problems needed to be overcome:

1: Merging input streams so that a single thread can do a blocking read on all workers' outputs
2: Waiting on multiple processes to die on a single thread. C has wait(), but I was surprised not to be able to find anything similar in Java. Any of the abstractions proposed on threads I read online ultimately used threads to wait on individual processes. The only solution I came up with for this was to busy-wait on the processes, using a timer to periodically interrupt the wait and iterate the process list.

The threaded solution seemed much cleaner for the small overhead involved.

The dead-workers data structure, on the other hand, I will move into the supervisor-state map.

nathanmarz commented:
I really don't like the threaded solution. Threads are a resource that if they fail to be properly cleaned up can cause weird issues to the supervisor. I think your proposal #2 is the way to go.

The way it's implemented now the logging will go to the supervisor log file, which definitely is not what we want. Can we get the logging to go to a separate worker log file for each worker? I think this will require controlling the classname the log messages are under and then an update to the logback config. 

The dead-workers data structure seems to be an optimization that's independent from redirecting the worker output. Can we remove all of that from here and re-open it as its own pull request once we finish this?

knusbaum commented:
I've removed the dead-workers optimization entirely and am working on the rest, although instead of a separate thread as I initially suggested, I'm currently checking the workers' output/status in supervisor/sync-processes (an option I overlooked).
It can be easily moved into its own thread if you want, but it doesn't seem necessary to me.

One issue: 
Without threads, I need a bit of state to keep track of the Process objects in order to check the exit code, but Process objects are not serializable.
Since the objects are meaningless to another instance of the supervisor anyway, it seems reasonable to track the objects outside the supervisor-state. What do you think?

knusbaum commented:
Working on logging into the workers' log files, but it doesn't look like a simple change of the logback config along with controlling the classname. I don't see any way of dynamically choosing a log file except with JoranConfigurator, which is a bit ugly.

knusbaum commented on storm-core/src/clj/backtype/storm/util.clj:
```diff
@@ -417,6 +419,22 @@
+    (fn []
+      (if log-fn
+        (while (.ready reader)
+          (log-fn (.readLine reader))))
```
I'll fix this. It's not non-blocking.

knusbaum commented:
I'll argue against trying to get the output of dying worker processes into the individual worker logs for two reasons:

1: Standard output from workers tends to occur when the JVM fails to launch due to improper supervisor configuration, not when there's something wrong with the worker.
2: This is a rare occurrence, and there doesn't appear to be an idiomatic way to report these errors to the workers' logs.

nathanmarz commented:
Standard output could also occur because someone was doing some debugging locally and accidentally deploys the topology to the cluster with stdout code. With logback we can control the size of the log files and prevent topologies like this from sucking up all the disk space on the cluster (and thereby causing a cascading failure). This is an important case to cover and not rare enough to ignore.

d2r commented:
It's possible to hook the worker's stdout/stderr to its logger, once the logger has been initialized.

Before the workers logger is initialized, it's unclear to me how to accomplish this.

If the supervisor simply opens and appends to the appropriate worker log directly, then it needs to be aware of the log file's path.  That means, I think, the supervisor needs to read the logging configuration and evaluate the template plugging in the correct port.  This also has the drawback that we miss other appenders configured, e.g., to send worker logs to a remote machine.

Or, if the supervisor initializes another logger for the worker that is launching, then it seems we could not use the same logging configuration, as setting the port in the logging system would necessarily affect the logging system across all worker launches.  I could be wrong about this, but I thought the logging context was static even across all threads.

Could there be a solution where we log the worker stdout/stderr to the supervisor log only until the worker initializes its own logger and can redirect its own stdout/stderr itself?

harshach committed:
STORM-1030. Hive Connector Fixes.

hmcl committed:
STORM-822: Kafka Spout New Consumer API

hmcl committed:
STORM-822: Kafka Spout New Consumer API

satishd committed:
STORM-676 addressed review comments

satishd committed:
STORM-676 Addressed review comments on API aligning with core window API

Commit:
port builtin-metrics to java

Commit:
Merge branch 'master' of https://github.com/srdo/storm into STORM-1549

STORM-1549: Add support for resetting tuple timeout from bolts via the OutputCollector

Commit:
Added STORM-1549 to Changelog

Commit:
Merge branch 'port-scheduler-to-java' of https://github.com/vesense/storm into STORM-1232

STORM-1232: port backtype.storm.scheduler.DefaultScheduler to java
STORM-1231: port backtype.storm.scheduler.EvenScheduler to java

Commit:
Added STORM-1232 and STORM-1231 to Changelog

Commit:
[STORM-1634] - 	Minor Refactoring of Resource Aware Scheduler

knusbaum committed:
Addressing Comments.

knusbaum committed:
adding code documentation explaining math of combining component resources.

Commit:
Merge branch 'maven-status' of https://github.com/vesense/storm into STORM-1624

Commit:
Added STORM-1624 to CHANGELOG.

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
Merge branch 'master' into pacemaker

hustfxj committed:
update code based on redsanket

Commit:
Merge branch 'nimbus-bug' of https://github.com/hustfxj/storm

Commit:
Added STORM-1623 to CHANGELOG

Commit:
[STORM-1636] - Supervisor shutdown with worker id pass in being nil

HeartSaVioR committed:
STORM-1602 Blobstore UTs are failed on Windows

* ensures objects of InputStream / OutputStream are closed after using
  * clojure: with-open
  * java: try-with-resource
* skip checking symbolic link in LocalizerTest when on Windows
  * Windows seems not handle symbolic link in compressed file properly

vesense committed:
STORM-1573: Add batch support for MongoInsertBolt

knusbaum committed:
Addressing comments and adding a bit more documentation.

knusbaum committed:
Merge branch 'Trident-RAS-v2' of https://github.com/knusbaum/incubator-storm

knusbaum committed:
Adding STORM-1616 to CHANGELOG.md

hustfxj committed:
add jmx for Pacemaker

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
Merge branch 'master' into pacemaker

Commit:
[STORM-1300] port backtype.storm.scheduler.resource-aware-scheduler-test to java.

Commit:
Minor

Commit:
Delete the clj code

hustfxj committed:
update PacemakerTest

Commit:
STORM-1617: Release Specific Documentation 0.9.x

Conflicts:
	.gitignore
	docs/README.md
	docs/_config.yml
	docs/_includes/footer.html
	docs/_includes/head.html
	docs/_includes/header.html
	docs/_layouts/about.html
	docs/_layouts/default.html
	docs/_layouts/documentation.html
	docs/_layouts/page.html
	docs/_layouts/post.html
	docs/assets/css/bootstrap.css
	docs/assets/css/bootstrap.css.map
	docs/assets/js/bootstrap.min.js
	docs/images/logos/alibaba.jpg
	docs/images/logos/groupon.jpg
	docs/images/logos/parc.png
	docs/images/logos/webmd.jpg
	docs/images/topology.png

Conflicts:
	.gitignore
	docs/README.md
	docs/STORM-UI-REST-API.md
	docs/_config.yml
	docs/_includes/footer.html
	docs/_includes/head.html
	docs/_includes/header.html
	docs/_layouts/about.html
	docs/_layouts/default.html
	docs/_layouts/documentation.html
	docs/_layouts/page.html
	docs/_layouts/post.html
	docs/assets/css/bootstrap.css
	docs/assets/css/bootstrap.css.map
	docs/assets/js/bootstrap.min.js
	docs/images/logos/alibaba.jpg
	docs/images/logos/groupon.jpg
	docs/images/logos/parc.png
	docs/images/logos/webmd.jpg

Commit:
STORM-1617: 0.10.x release docs

Conflicts:
	docs/images/topology.png

Commit:
STORM-1617: added in 1.x specific doc changes

Commit:
STORM-1617: Added in 2.x specific doc changes

HeartSaVioR committed:
Merge branch 'STORM-1629'

HeartSaVioR committed:
add STORM-1629 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1602'

HeartSaVioR committed:
add STORM-1602 to CHANGELOG.md

Commit:
STORM-1625: Move storm-sql dependencies out of lib folder

brymaven committed:
Correct typo on example javadoc

arunmahadevan committed:
[STORM-1585] Address review comments

Commit:
STORM-1464: storm-hdfs support for multiple output files and partitioning

Commit:
Revised wipeZeros() in SlotBasedCounter to avoid creating new hash set.

hmcl committed:
STORM-822: Kafka Spout New Consumer API
 - Refactored code to avoid keeping records data inside spout state
 - Refactored code to specify output fields per stream and build tuples per topic
 - Implement exponential backoff retry strategy
 - Send one tuple per call to nextTuple

hmcl committed:
STORM-822: Kafka Spout New Consumer API
 - Refactored code to avoid keeping records data inside spout state
 - Refactored code to specify output fields per stream and build tuples per topic
 - Implement exponential backoff retry strategy
 - Send one tuple per call to nextTuple

srdo committed:
Made StormTimer join task thread on close

Commit:
Merge branch 'versioned-docs' of https://github.com/revans2/incubator-storm into STORM-1617

STORM-1617: Versioned Docs

Commit:
Merge branch 'STORM-1636' of https://github.com/jerrypeng/storm into STORM-1636

STORM-1636: Supervisor shutdown with worker id pass in being nil

Commit:
Added STORM-1636 to Changelog

Commit:
Merge branch 'metricsFile' of https://github.com/DigitalPebble/storm into STORM-584

STORM-584: Fix logging for LoggingMetricsConsumer metrics.log file

Commit:
Added STORM-584 to Changelog

Commit:
Merge branch 'STORM-1631' of https://github.com/jerrypeng/storm into STORM-1631

STORM-1631: Storm CGroup bugs

Commit:
Added STORM-1631 to Changelog

Commit:
Merge branch 'STORM-1614_backpressure_init_and_cleanup_changes' of https://github.com/abellina/storm into STORM-1614

STORM-1614: backpressure changes and test

Commit:
STORM-1648: drpc spout reconnect on failure

Commit:
Merge branch 'STORM-1556' of https://github.com/wuchong/storm into STORM-1556

STORM-1556: nimbus.clj/wait-for-desired-code-replication wrong reset for current-replication-count-jar in local mode

Commit:
Added STORM-1556 to Changelog

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
Merge branch 'master' into pacemaker

hustfxj committed:
Coda Hale for the stats

hustfxj committed:
revert the StormMetricsRegistry

hustfxj committed:
Merge branch 'master' into supervisor and update supervisor based STORM-1631&STORM-1636

satishd committed:
STORM-676 Addressed review comments from Arun

HeartSaVioR committed:
STORM-1630 "create symbolic link" needs elevation or setting privilege about creating symbolic link

* Add document to guide how to setup account to be able to 'create symbolic link' on Windows

satishd committed:
STORM-676 Upmerged and resolved conflicts

satishd committed:
STORM-676 Addressed review comments

satishd committed:
STORM-676 Refactoring of WindowConfig APIs

hustfxj committed:
update

knusbaum committed:
Merge branch 'STORM-1613' of https://github.com/satishd/storm

HeartSaVioR committed:
STORM-1654 HBaseBolt creates tick tuples with no interval when we don't set flushIntervalSecs

* set 'default' flush interval seconds (1s) to HBaseBolt
  * since taking half of message timeout secs doesn't work

HeartSaVioR committed:
STORM-1655 Flux doesn't set return code to non-zero when there's any exception while deploying topology to remote cluster

* Modify Flux.runCli() to not catching exceptions so that exception can be propagated to main()

Commit:
STORM-1537: Upgrade to kryo3 in master

Commit:
Merge branch 'STORM-1648_drpcspout_reconnect_in_fail' of https://github.com/abellina/storm into STORM-1648

STORM-1648: drpc spout reconnect on failure

Commit:
Added STORM-1648 to Changelog.

Commit:
Merge branch 'STORM-1268' of https://github.com/unsleepy22/storm into STORM-1268

STORM-1268: port builtin-metrics to java

Commit:
Added STORM-1268 to Changelog

Commit:
Merge branch 'storm-sql' of https://github.com/abhishekagarwal87/storm into STORM-1625

STORM-1625: Move storm-sql dependencies out of lib folder

knusbaum committed:
Adding documentation for Trident RAS API

Commit:
Added STORM-1625 to Changelog

brymaven committed:
Fix indent on clojure testing function

knusbaum committed:
Merge branch 'fix-timer-shutdown-race' of https://github.com/srdo/storm into asf-master

brymaven committed:
Add Clojure examples for rolling top words and exclamation

knusbaum committed:
Merge branch 'master' of https://github.com/hukun01/storm

knusbaum committed:
Merge branch 'private_constructor' of https://github.com/lujinhong/storm

Commit:
Merge two functions

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
meter/histogram/Gauge

hustfxj committed:
update

hustfxj committed:
Merge branch 'master' into supervisor

Commit:
Merge branch 'kryo3-master' of http://github.com/abhishekagarwal87/storm into STORM-1537

Commit:
Added STORM-1537 to CHANGELOG.

Commit:
Merge branch 'STORM-1654' of http://github.com/HeartSavioR/storm into STORM-1654

Commit:
Added STORM-1654 to CHANGELOG.

Commit:
Added STORM-1654 to CHANGELOG.

hustfxj committed:
Update SyncSupervisorEvent.java

Commit:
STORM-1271: Port backtype.storm.daemon.task to java

HeartSaVioR committed:
Merge branch 'STORM-1655'

HeartSaVioR committed:
add STORM-1655 to CHANGELOG.md

Commit:
Modify comments

Commit:
Modify comment

Commit:
Clarified docs.

Commit:
Minor

HeartSaVioR committed:
Merge branch 'STORM-1630'

HeartSaVioR committed:
add STORM-1630 to CHANGELOG.md

satishd committed:
Merge remote-tracking branch 'upstream/master' into STORM-676

satishd committed:
STORM-676 Refactored HBase#put APIs for hbase version upgrade.

vesense committed:
STORM-1659:Add documents for external projects

vesense committed:
refactor gitignore

HeartSaVioR committed:
Merge branch 'STORM-1659-DOCS' of https://github.com/vesense/storm into STORM-1659

vesense committed:
remove flux gitignore

Commit:
Merge branch 'STORM-676' of http://github.com/satishd/storm into STORM-676

Commit:
Added STORM-676 to CHANGELOG.

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

satishd committed:
STORM-1652 Added trident windowing API documentation to Trident API doc.

arunmahadevan committed:
[STORM-1544] Document Debug/Sampling of Topologies

Added documentation for logging/sampling events in a topology.

knusbaum committed:
Minor fix to RAS doc.

Update Trident-RAS-API.md

knusbaum committed:
Merge branch 'trident-doc'

knusbaum committed:
Fixing link.

knusbaum committed:
Fixing bullet-points for jekyll

Commit:
Updated README to have correct copy command

Commit:
Updated javadocs to not have timestamps

Commit:
Missed a place for notimestamp

Commit:
Merge branch '1300' of https://github.com/zhuoliu/storm into STORM-1300

satishd committed:
STORM-1649 kryo serialization in windowing

satishd committed:
STORM-1649 kryo serialization in windowing

satishd committed:
STORM-1652 Addressed review comments

HeartSaVioR committed:
Merge branch 'STORM-1030-V1' of https://github.com/harshach/incubator-storm into STORM-1030

HeartSaVioR committed:
add STORM-1030 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1030'

Commit:
adding a unit test

Commit:
Merge branch 'ras_refactor' of https://github.com/jerrypeng/storm into STORM-1634

STORM-1634: Refactoring of Resource Aware Scheduler

Commit:
Added STORM-1634 to Changelog

Longda-Feng committed:
Merge branch 'pacemaker' of https://github.com/hustfxj/storm into STORM-1611

Longda-Feng committed:
Update [STORM-1611] into CHANGELING

Longda-Feng committed:
Merge branch 'pacemaker-state-factory-test' of https://github.com/hustfxj/storm into STORM-1610

Longda-Feng committed:
update STORM-1610 PR into CHANGELOG

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
Merge branch 'master' into supervisor

Commit:
STORM-1663: Stats couldn't handle null worker HB.

Commit:
STORM-1665: Worker cannot instantiate kryo

Commit:
STORM-1666: Kill from the UI fails silently

ptgoetz committed:
Merge branch 'STORM-1660' of github.com:vesense/storm

ptgoetz committed:
add STORM-1660 to changelog

priyank5485 committed:
STORM-1668: Fix silent failing of flux for setting a non-existent property.

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
update supervisor based on revans2 and longdafeng

vesense committed:
STORM-1669: Fix SolrUpdateBolt flush bug

HeartSaVioR committed:
Merge branch 'STORM-1573' of https://github.com/vesense/storm into STORM-1573

HeartSaVioR committed:
add STORM-1573 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1652' of https://github.com/satishd/storm into STORM-1652

Commit:
Merge branch 'STORM-1663' of https://github.com/revans2/incubator-storm into STORM-1663

STORM-1663: Stats couldn't handle null worker HB.
STORM-1665: Worker cannot instantiate kryo
STORM-1666: Kill from the UI fails silently

Commit:
Added STORM-1663 STORM-1665 and STORM-1666 to Changelog

Commit:
Added STORM-822 to Changelog

Commit:
[STORM-1671] Enable logviewer to delete a dir without yaml

Commit:
Merge branch 'STORM-822' of https://github.com/hmcl/storm-apache

ptgoetz committed:
Merge branch 'STORM-1649' of github.com:satishd/storm

ptgoetz committed:
add STORM-1649 to changelog

Commit:
Merge branch 'task' of https://github.com/abhishekagarwal87/storm into STORM-1271

STORM-1271: Port backtype.storm.daemon.task to java

priyank5485 committed:
STORM-1535: Make sure hdfs key tab login happens only once for multiple bolts/executors.

Commit:
Added STORM-1271 to Changelog

ptgoetz committed:
Revert "STORM-1649 kryo serialization in windowing"

This reverts commit 576499baba14926a564a51e0018f3ead86809ab1.

ptgoetz committed:
Revert "add STORM-1649 to changelog"

This reverts commit 223b615d1bea665b66c5c0fe8fd89cdc2e6dd94d.

hustfxj committed:
Merge branch 'master' of github.com:apache/storm

hustfxj committed:
fix about RunProfilerActions

brymaven committed:
Clean up macros in storm starter

knusbaum committed:
Merge branch 'STORM-1668' of https://github.com/priyank5485/storm

knusbaum committed:
Adding STORM-1668 to CHANGELOG.md

HeartSaVioR committed:
STORM-1673 log4j2/worker.xml refers old package of LoggerMetricsConsumer

* fix missing package changes

arunmahadevan committed:
Updating the doc to reflect the new default behavior

ptgoetz committed:
Merge branch 'STORM-1669' of github.com:vesense/storm

ptgoetz committed:
add STORM-1669 to changelog

ptgoetz committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/storm

satishd committed:
STORM-1670 version store should consider ony active/finished version files

Commit:
Merge branch 'supervisor' of https://github.com/hustfxj/storm into STORM-1279

STORM-1279: port backtype.storm.daemon.supervisor to java

Commit:
Added STORM-1279 to Changelog

ptgoetz committed:
Merge branch 'STORM-1673' of https://github.com/HeartSaVioR/storm

ptgoetz committed:
add STORM-1673 to changelog

ptgoetz committed:
Merge branch '1671' of github.com:zhuoliu/storm

ptgoetz committed:
add STORM-1671 to changelog

satishd committed:
STORM-1670 Removed DirectoryStream changes

Commit:
Merge branch '1671' of https://github.com/zhuoliu/storm into STORM-1671

ptgoetz committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/storm

satishd committed:
STORM-1670 Fixed unit tests

HeartSaVioR committed:
STORM-1677 Test resource files are excluded from source distribution, which makes logviewer-test failing

* rename '.log' files to '.log.test' which can avoid exclusion when building source distribution

vesense committed:
add scheduler docs

Commit:
[STORM-1681] - Bug in scheduling cyclic topologies when scheduling with RAS

knusbaum committed:
Merge branch 'STORM-515' of https://github.com/brymaven/storm

knusbaum committed:
Adding STORM-515 to the changelog.

Commit:
only check non-system streams by default

Commit:
STORM-1680, Added Kafka Spout Config FetchByte to Storm-Kafka

ptgoetz committed:
Merge branch 'STORM-1677' of github.com:HeartSaVioR/storm

ptgoetz committed:
add STORM-1677 to changelog

ptgoetz committed:
Merge branch 'STORM-1670' of github.com:satishd/storm

ptgoetz committed:
add STORM-1670 to changelog

Commit:
[STORM-1286] port kill_workers to java.

Commit:
[STORM-1687] divide by zero in StatsUtil

Commit:
Merge branch '1286' of https://github.com/zhuoliu/storm into STORM-1286

vesense committed:
fix documents 404

vesense committed:
fix nimbus ha description

Commit:
Merge branch 'STORM-1494' of https://github.com/dossett/storm into STORM-1464

Commit:
Add 1464 to CHANGELOG

vesense committed:
minor fix

HeartSaVioR committed:
STORM-1704 When logviewer_search.html opens daemon file, next search always show no result

* ensures that is-daemon parameter is passed around multiple searches
* set logviewerUrl to '/daemonlog' when search is done with is-daemon=yes

Commit:
Merge branch 'STORM-1687' of https://github.com/zhuoliu/storm into STORM-1687

STORM-1687: divide by zero in StatsUtil

Commit:
added STORM-1687 to Changelog.

Commit:
Merge branch 'STORM-1658' of https://github.com/vesense/storm into STROM-1658

STORM-1658: documents improvements and links fixes

Commit:
STORM-1693: Move stats cleanup to executor shutdown

Commit:
[STORM-1696] status not sync if zk fails in backpressure

Commit:
minor

Commit:
Merge branch 'STORM-1681' of https://github.com/jerrypeng/storm

Commit:
added STORM-1681 to Changelog

Commit:
Merge branch 'STORM-1585' of https://github.com/arunmahadevan/storm into STORM-1585

Commit:
Added STORM-1585 to CHANGELOG.

HeartSaVioR committed:
Merge branch 'STORM-1693-master' of https://github.com/abhishekagarwal87/storm into STORM-1693

HeartSaVioR committed:
add STORM-1693 to CHANGELOG.md

HeartSaVioR committed:
Merge branch '1696b' of https://github.com/zhuoliu/storm into STORM-1696

HeartSaVioR committed:
add STORM-1696 to CHANGELOG.md

arunmahadevan committed:
[STORM-1714] StatefulBolts ends up as normal bolts while using TopologyBuilder.setBolt without parallelism

mjsax committed:
minor: fixed broken link

arunmahadevan committed:
[STORM-1714] refactored common logic into BaseStatefulBoltExecutor

HeartSaVioR committed:
Merge branch 'STORM-1649-1' of https://github.com/satishd/storm into STORM-1649

HeartSaVioR committed:
add STORM-1649 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1680'

HeartSaVioR committed:
add STORM-1680 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'storm-1683-ui-viz-uncheck-sys-streams-master' of https://github.com/d2r/storm into STORM-1683

HeartSaVioR committed:
add STORM-1683 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1714' of https://github.com/arunmahadevan/storm into STORM-1714

HeartSaVioR committed:
add STORM-1714 to CHANGELOG.md

HeartSaVioR committed:
Merge branch 'STORM-1704'

HeartSaVioR committed:
add STORM-1704 to CHANGELOG.md

priyank5485 committed:
STORM-1535: Removing the unnecessary comment.

ptgoetz committed:
Merge branch 'STORM-1679' of github.com:vesense/storm

ptgoetz committed:
Add STORM-1679 to changelog

ptgoetz committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/storm

ptgoetz committed:
Merge branch 'STORM-1544' of github.com:arunmahadevan/storm

ptgoetz committed:
add STORM-1544 to changelog

Commit:
Flush stdout and stderr before calling "os.execvp" to prevent log loss.

ptgoetz committed:
Merge branch 'STORM-1535' of github.com:priyank5485/storm

ptgoetz committed:
add STORM-1535 to changelog

mjsax committed:
minor: improved documenation layout

This closes #1363

ziyunhx committed:
update the java version.

Storm drops supporting JDK 1.6 in Apache Storm 0.10.0 and above.

arunmahadevan committed:
Merge branch 'flush_launch_command_to_stderr' of https://github.com/dsKarthick/storm-1 into STORM-1733

arunmahadevan committed:
Added STORM-1733 to CHANGELOG.md

ptgoetz committed:
Merge branch 'master' of github.com:ziyunhx/storm

ptgoetz committed:
add STORM-1739 to changelog

knusbaum committed:
Test Pacemaker Fix

knusbaum committed:
Bit of cleanup
